pkg/resiliency/error.go

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package resiliency
+
+import (
+	"fmt"
+)
+
+const (
+	none Tag = ""
+
+	// ResExt tracks unavailability of external resource (malicious or accidental)
+	ResExt Tag = "resExt"
+
+	// ResLimit tracks constraints on resources or API
+	ResLimit Tag = "resLimit"
+
+	// ResUnavailable tracks external manipulation (manual or accidental).
+	ResUnavailable Tag = "resUnavailable"
+
+	// ResDatapath tracks datapath misbehavior.
+	ResDatapath Tag = "resDatapath"
+
+	// ResControlplane tracks control plane misbehavior.
+	ResControlplane Tag = "resControlplane"
+
+	// ExtSys tracks failure of system interactions.
+	ExtSys Tag = "extSys"
+)
+
+type (
+	// Tag represent a resiliency issue category.
+	Tag string
+
+	taggedErr struct {
+		error
+		tag Tag
+	}
+
+	// RetryableErr tracks error types that could be retried.
+	RetryableErr struct {
+		taggedErr
+	}
+)
+
+// NewErr returns a new instance.
+func NewErr(e error, t Tag) taggedErr {
+	return taggedErr{
+		error: e,
+		tag:   t,
+	}
+}
+
+// Error returns a human representation.
+// TODO do we want to decorate at all??
+func (r taggedErr) Error() string {
+	if r.tag == none {
+		return r.error.Error()
+	}
+
+	return fmt.Sprintf("[%s] %s", r.tag, r.error.Error())
```

Comment:
I wonder if this should actually be a separate attribute in structured logging rather than embedded in the message. I could see a way for that to be filtered by users to then evaluate impact and prioritize which types of errors to do something about.

Reply:
Thank you for the input @joestringer!

Not sure?? I was thinking that this would help categorizing errors using tags in conjunction with error codes (later??) 
ie `[resExt: 1001] K8s api server unreachable.`   
We could then document codes with further description, potential resolves, ... to help user get to root cause based on these tag/code combo??
 
It seems best if this tag:code goes along with the error at the point its raises. More often than not, we are going to have multiple errors bubble up to the call site. The current thought is to leverage errors.Join to compile the list (which basically dumps the errors separated with new lines). Hence a single log line could reveal several errors with different tags/codes. Does this make sense?

Reply:
I see, for that style use case I can see why integrating it into the message could be helpful. What I was pondering is if a user might want to filter all errors of a specific type like `resExt` in some external system like splunk, elastic, etc.; having a dedicated field would help in that case. But I guess both points are moot until we widely use this form to present errors to the user.

Reply:
I see. I'm not super strongly opinionated in the above, but thought it's worth raising a discussion either way. At this point the PR doesn't have the related changes, so I think we can resolve out this conversation. If/when we revisit the proposal from this conversation thread, we can continue the discussion.